services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java

```diff
@@ -1048,12 +1049,27 @@ private void migrateFederatedIdentityId(BrokeredIdentityContext context, UserMod
 
     private void updateToken(BrokeredIdentityContext context, UserModel federatedUser, FederatedIdentityModel federatedIdentityModel) {
         if (context.getIdpConfig().isStoreToken() && !ObjectUtil.isEqualOrBothNull(context.getToken(), federatedIdentityModel.getToken())) {
-            federatedIdentityModel.setToken(context.getToken());
+            try {
+                // like in OIDCIdentityProvider.exchangeStoredToken()
+                // we shouldn't override the refresh token if it is null in the context and not null in the DB
+                // as for google IDP it will be lost forever
+                AccessTokenResponse previousResponse = JsonSerialization.readValue(federatedIdentityModel.getToken(), AccessTokenResponse.class);
```

Comment:
@geoffreyfourmis Looks like we need to check first if the response is stored prior to running your code. That should fix the test failures.

Reply:
Thanks, should be better